internal/manifest/oci_index.go

```diff
@@ -73,39 +84,89 @@ func (index *OCI1IndexPublic) UpdateInstances(updates []ListUpdate) error {
+		return true
+	}
+	return false
+}
+
+type instanceCandidate struct {
+	platformIndex    int           // Index of the candidate in platform.WantedPlatforms: lower numbers are preferred; or math.maxInt if the candidate doesn’t have a platform
+	isZstd           bool          // tells if particular instance if zstd instance
+	manifestPosition int           // A zero-based index of the instance in the manifest list
+	digest           digest.Digest // Instance digest
+}
+
+func (ic instanceCandidate) isPreferredOver(other *instanceCandidate, preferGzip bool) bool {
+	switch {
+	case ic.platformIndex != other.platformIndex:
+		return ic.platformIndex < other.platformIndex
+	case ic.isZstd != other.isZstd:
+		if !preferGzip {
+			return ic.isZstd
+		} else {
+			return !ic.isZstd
+		}
+	case ic.manifestPosition != other.manifestPosition:
+		return ic.manifestPosition < other.manifestPosition
+	}
+	panic("internal error: invalid comparision between two candidates") // This should not be reachable because in all calls we make, the two candidates differ at least in manifestPosition.
+}
+
+// chooseInstance is a private equivalent to ChooseInstanceByCompression,
+// shared by ChooseInstance and ChooseInstanceByCompression.
+func (index *OCI1IndexPublic) chooseInstance(ctx *types.SystemContext, preferGzip types.OptionalBool) (digest.Digest, error) {
+	didPreferGzip := false
+	if preferGzip == types.OptionalBoolTrue {
+		didPreferGzip = true
+	}
 	wantedPlatforms, err := platform.WantedPlatforms(ctx)
 	if err != nil {
 		return "", fmt.Errorf("getting platform information %#v: %w", ctx, err)
 	}
-	for _, wantedPlatform := range wantedPlatforms {
-		for _, d := range index.Manifests {
-			if d.Platform == nil {
-				continue
+	var bestMatch *instanceCandidate
+	bestMatch = nil
+	for manifestIndex, d := range index.Manifests {
+		candidate := instanceCandidate{platformIndex: math.MaxInt, manifestPosition: manifestIndex, isZstd: instanceIsZstd(d), digest: d.Digest}
+		if d.Platform != nil {
+			for platformIndex, wantedPlatform := range wantedPlatforms {
+				imagePlatform := imgspecv1.Platform{
+					Architecture: d.Platform.Architecture,
+					OS:           d.Platform.OS,
+					OSVersion:    d.Platform.OSVersion,
+					OSFeatures:   slices.Clone(d.Platform.OSFeatures),
+					Variant:      d.Platform.Variant,
+				}
+				if platform.MatchesPlatform(imagePlatform, wantedPlatform) {
+					candidate.platformIndex = platformIndex
+					break
+				}
 			}
-			imagePlatform := imgspecv1.Platform{
-				Architecture: d.Platform.Architecture,
-				OS:           d.Platform.OS,
-				OSVersion:    d.Platform.OSVersion,
-				OSFeatures:   slices.Clone(d.Platform.OSFeatures),
-				Variant:      d.Platform.Variant,
-			}
-			if platform.MatchesPlatform(imagePlatform, wantedPlatform) {
-				return d.Digest, nil
+		}
+		if d.Platform == nil || candidate.platformIndex != math.MaxInt {
```

Comment:
This works, but it was not obvious to me without your comment elsewhere alluding to this.

This is repetitive for the `d.Platform` case, and _just far enough_ from the `wantedPlatform` loop for me to miss the connection with that loop.

Maybe
```go
if d.Platform != nil {
  for … { … }
  if candidate.platformIndex == math.MaxInt { // No platform match found
    continue
  }
}
```
or all the way to full formal tracking:
```go
if d.Platform != nil {
  foundPlatform := false 
  for … { if … { foundPlatform = true; candidate.platformIndex = …; break } }
  if !foundPlatform {
    continue
  }
}
```
in both cases, I think this `if` statement can be removed afterwards.